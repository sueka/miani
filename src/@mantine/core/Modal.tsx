import { useDraggable, useDroppable } from '@dnd-kit/core'
import { Modal as OrigModal, type ModalProps } from '@mantine/core'
import { useLayoutEffect, useMemo, useState } from 'react'
import { tss } from 'tss-react'
import { v4 } from 'uuid'
import assert from '../../lib/assert'

interface Inset {
  top: number
  left: number
}

const useDraggableStyles = tss
  .withParams<{
    dragging: Inset | null
  }>()
  .create(({ dragging }) => ({
    content:
      dragging !== null
        ? {
            position: 'relative',
            top: `${dragging.top}px`,
            left: `${dragging.left}px`,
          }
        : {},
  }))

const Modal: React.FC<ModalProps> = ({
  opened,
  onClose,
  title,
  children,
  ...props
}) => {
  const id = useMemo(v4, [])
  const [dragging, setDragging] = useState<Inset | null>(null)

  const {
    setNodeRef: draggable,
    setActivatorNodeRef: handler,
    active,
    listeners,
    attributes,
  } = useDraggable({
    id,
  })

  const { setNodeRef: droppable } = useDroppable({
    id,
  })

  const { classes: draggableClasses } = useDraggableStyles({
    dragging,
  })

  useLayoutEffect(() => {
    // dragging
    if (active !== null) {
      const { initial, translated } = active.rect.current

      assert(initial !== null)
      assert(translated !== null)

      if (active.id === id) {
        setDragging({
          top: translated.top - initial.top,
          left: translated.left - initial.left,
        })
      }
    }
  }, [active, active?.rect.current])

  return (
    <OrigModal.Root
      opened={opened}
      onClose={onClose}
      ref={droppable}
      {...props}
    >
      <OrigModal.Overlay />
      <OrigModal.Content
        ref={draggable}
        classNames={draggableClasses}
        {...attributes}
      >
        <OrigModal.Header ref={handler} {...listeners}>
          <OrigModal.Title>{title}</OrigModal.Title>
          <OrigModal.CloseButton data-no-dnd />
        </OrigModal.Header>
        <OrigModal.Body>{children}</OrigModal.Body>
      </OrigModal.Content>
    </OrigModal.Root>
  )
}

export default Modal
