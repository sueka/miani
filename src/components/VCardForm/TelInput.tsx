import { useSortable } from '@dnd-kit/sortable'
import { CSS } from '@dnd-kit/utilities'
import {
  ActionIcon,
  Checkbox,
  Group,
  Input,
  TextInput,
  ThemeIcon,
} from '@mantine/core'
import { useValidatedState } from '@mantine/hooks'
import { IconBackspace, IconGripVertical } from '@tabler/icons-react'
import { useLayoutEffect } from 'react'
import { useIntl } from 'react-intl'
import { useRecoilState, useResetRecoilState, useSetRecoilState } from 'recoil'
import nes from '../../lib/nes'
import r from '../../lib/tags/r'
import { phoneNumberValue } from '../../patterns'
import sharedState from '../../recoil/atoms/sharedState'
import telIdsState from '../../recoil/atoms/tel/telIdsState'
import telState from '../../recoil/atoms/tel/telState'

interface Props {
  telId: string
}

const TelInput: React.FC<Props> = ({ telId }) => {
  const { formatMessage } = useIntl()
  const [recoilTel, setRecoilTel] = useRecoilState(telState(telId))
  const [shared, setShared] = useRecoilState(sharedState(telState(telId).key))
  const resetTel = useResetRecoilState(telState(telId))
  const setTelIds = useSetRecoilState(telIdsState)

  const [tel, setTel] = useValidatedState<string | null>(
    recoilTel,
    (value) =>
      value === null ||
      // value === '' ||
      r`^${phoneNumberValue}$`.test(value),
  )

  useLayoutEffect(() => {
    setTel(recoilTel)
  }, [setTel, recoilTel])

  const {
    attributes,
    listeners,
    setNodeRef: draggable,
    setActivatorNodeRef: handle,
    transform,
    transition,
  } = useSortable({ id: telId })

  return (
    <Input.Wrapper
      ref={draggable}
      style={{
        transform: CSS.Transform.toString(transform),
        transition,
      }}
      {...attributes}
      error={
        !tel.valid && (
          <span
            dangerouslySetInnerHTML={{
              __html: formatMessage({
                defaultMessage:
                  'Should be a <i>phone-number-value</i> on p. 37, RFC 2426.',
              }),
            }}
          />
        )
      }
    >
      <Group gap="xs">
        <Checkbox
          checked={shared}
          onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
            setShared(event.currentTarget.checked)
          }}
        />
        <TextInput
          flex={1}
          // NOTE: This phone number does not exist because "3" (and "6") are exclusive area codes and local codes do not start with "0" in Japan (+81).
          placeholder="+81-3-0123-4567"
          value={tel.value ?? ''}
          onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
            setTel(nes(event.currentTarget.value))
            setRecoilTel(nes(event.currentTarget.value))
          }}
          error={!tel.valid}
        />
        <ActionIcon
          variant="light"
          onClick={() => {
            if (
              (tel.value ?? '').match(/^\s*$/) ||
              confirm(
                formatMessage(
                  { defaultMessage: 'Delete the phone number: {tel}' },
                  { tel: tel.value },
                ),
              )
            ) {
              resetTel()
              setTelIds((telIds) => telIds.filter((t) => t !== telId))
            }
          }}
        >
          <IconBackspace />
        </ActionIcon>
        <ThemeIcon
          color="gray"
          variant="transparent"
          ref={handle}
          {...listeners}
          style={{
            touchAction: 'none',
          }}
        >
          <IconGripVertical />
        </ThemeIcon>
      </Group>
    </Input.Wrapper>
  )
}

export default TelInput
