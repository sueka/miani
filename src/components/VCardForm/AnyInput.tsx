import { Checkbox, Group, Input, TextInput } from '@mantine/core'
import { useMemo } from 'react'
import { useRecoilState } from 'recoil'
import { v4 } from 'uuid'
import anyState from '../../recoil/atoms/any/anyState'
import sharedState from '../../recoil/atoms/sharedState'

interface Props {
  type: string
}

const AnyInput: React.FC<Props> = ({ type }) => {
  const [any, setAny] = useRecoilState(anyState(type))
  const [shared, setShared] = useRecoilState(sharedState(anyState(type).key))
  const inputId = useMemo(v4, [])

  return (
    <Input.Wrapper label={type} labelProps={{ htmlFor: inputId }}>
      <Group gap="xs">
        <Checkbox
          checked={shared}
          onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
            setShared(event.currentTarget.checked)
          }}
        />
        <TextInput
          id={inputId}
          flex={1}
          value={any}
          onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
            setAny(event.currentTarget.value)
          }}
        />
      </Group>
    </Input.Wrapper>
  )
}

export default AnyInput
