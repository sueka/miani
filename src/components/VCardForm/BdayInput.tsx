import { Checkbox, Group, Input } from '@mantine/core'
import { DatePicker } from '@mantine/dates'
import React, { useState } from 'react'
import { useRecoilState } from 'recoil'

import toLegacyDate from '../../lib/Temporal/toLegacyDate'
import toPlainDate from '../../lib/Temporal/toPlainDate'
import noYearState from '../../recoil/atoms/bday/noYearState'
import bdayState from '../../recoil/atoms/bdayState'
import sharedState from '../../recoil/atoms/sharedState'

const BdayInput: React.FC = () => {
  const [bday, setBday] = useRecoilState(bdayState)
  const [date, setDate] = useState(bday !== null ? toLegacyDate(bday) : null)
  const [shared, setShared] = useRecoilState(sharedState(bdayState.key))
  const [noYear, setMonthDay] = useRecoilState(noYearState)

  return (
    <Input.Wrapper label="Birth date">
      <Group gap="sm">
        <Checkbox
          checked={shared}
          onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
            setShared(event.currentTarget.checked)
          }}
        />
        <DatePicker
          allowDeselect
          date={date ?? undefined}
          onDateChange={(date: Date | undefined) => {
            setDate(date ?? null)
          }}
          value={bday !== null ? toLegacyDate(bday) : null}
          onChange={(newBday: Date | null) => {
            setBday(newBday !== null ? toPlainDate(newBday) : null)
          }}
        />
      </Group>
      <Checkbox
        checked={noYear}
        onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
          setMonthDay(event.currentTarget.checked)
        }}
        label="Omits the year of birth of the object the vCard represents."
      />
    </Input.Wrapper>
  )
}

export default BdayInput
