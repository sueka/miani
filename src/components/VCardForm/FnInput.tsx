import { TextInput } from '@mantine/core'
import { useValidatedState } from '@mantine/hooks'
import React, { useLayoutEffect } from 'react'
import { FormattedMessage, useIntl } from 'react-intl'
import { useRecoilState, useRecoilValue } from 'recoil'

import exit from '../../lib/exit'
import r from '../../lib/tags/r'
import { text, textValue } from '../../patterns'
import fnState from '../../recoil/atoms/fnState'
import versionState from '../../recoil/atoms/vCard/versionState'

const FnInput: React.FC = () => {
  const { formatMessage } = useIntl()
  const version = useRecoilValue(versionState)
  const [recoilFn, setRecoilFn] = useRecoilState(fnState)

  const [fn, setFn] = useValidatedState<string>(recoilFn, (value) => {
    switch (version) {
      case '3.0':
        return r`^${textValue}$`.test(value)

      case '4.0':
        return r`^${text}$`.test(value)
    }
  })

  useLayoutEffect(() => {
    setFn(recoilFn)
  }, [setFn, recoilFn])

  return (
    <TextInput
      required
      label={<FormattedMessage defaultMessage="Formatted name" />}
      placeholder="Mr. John Q. Public\, Esq."
      autocomplete="name"
      value={fn.value}
      error={
        !fn.valid && (
          <span
            dangerouslySetInnerHTML={{
              __html:
                version === '3.0'
                  ? formatMessage({
                      defaultMessage:
                        'Should be a <i>text-value</i> on p. 37, RFC 2426.',
                    })
                  : version === '4.0'
                    ? formatMessage({
                        defaultMessage:
                          'Should be a <i>text</i> on p. 10, RFC 6350.',
                      })
                    : exit(),
            }}
          />
        )
      }
      onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
        setFn(event.currentTarget.value)
        setRecoilFn(event.currentTarget.value)
      }}
    />
  )
}

export default FnInput
