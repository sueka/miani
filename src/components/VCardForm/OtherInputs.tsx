import { Fieldset, Stack, TagsInput } from '@mantine/core'
import { FormattedMessage } from 'react-intl'
import { useRecoilState } from 'recoil'
import anyTypesState from '../../recoil/atoms/anyTypesState'
import AnyInput from './AnyInput'

const OtherInputs: React.FC = () => {
  const [anyTypes, setAnyTypes] = useRecoilState(anyTypesState)

  return (
    <Fieldset
      legend={<FormattedMessage defaultMessage="Others" />}
      px={{ base: 'md' }}
      pb={{ base: 'md' }}
    >
      <Stack gap="xs">
        <TagsInput value={anyTypes} onChange={setAnyTypes} />
        {anyTypes.map((type) => (
          <AnyInput {...{ type }} />
        ))}
      </Stack>
    </Fieldset>
  )
}

export default OtherInputs
