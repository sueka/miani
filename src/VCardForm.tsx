import AnyInput from './AnyInput'
import BdayPicker from './BdayPicker'
import FnInput from './FnInput'
import NInput from './NInput'

const VCardForm: React.FC = () => (
  <>
    <FnInput />
    <NInput />
    <BdayPicker />
    <AnyInput type="FOO" />
  </>
)

export default VCardForm
