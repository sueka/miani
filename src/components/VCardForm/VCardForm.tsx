import { Stack } from '@mantine/core'
import BdayInput from './BdayInput'
import FnInput from './FnInput'
import NInput from './NInput/NInput'
import OtherInputs from './OtherInputs'
import TelInputs from './TelInputs'
import XPhoneticNamesInput from './XPhoneticNamesInput'

const VCardForm: React.FC = () => (
  <Stack gap="xs">
    <FnInput />
    <NInput />
    <XPhoneticNamesInput />
    <BdayInput />
    <TelInputs />
    <OtherInputs />
  </Stack>
)

export default VCardForm
