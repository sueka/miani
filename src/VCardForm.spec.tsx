import { expect, setSystemTime, test } from 'bun:test'
import { MantineProvider } from '@mantine/core'
import { render } from 'react-dom'
import { RecoilRoot } from 'recoil'

import VCardForm from './VCardForm'

test('<VCardForm>', () => {
  setSystemTime(new Date('2006-01-02T15:04:05Z'))

  render(
    <MantineProvider>
      <RecoilRoot>
        <VCardForm />
      </RecoilRoot>
    </MantineProvider>,
    document.body,
  )

  expect(document.body.innerHTML).toMatchSnapshot()
})
