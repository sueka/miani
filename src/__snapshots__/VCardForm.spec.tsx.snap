// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`<VCardForm> 1`] = `"<style data-mantine-styles="classes">@media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}</style><style data-mantine-styles="inline">.__m__-P0-0{--grid-gutter:var(--mantine-spacing-md);}</style><div class="m-410352e9 mantine-Grid-root __m__-P0-0"><div class="m-dee7bd2f mantine-Grid-inner"><style data-mantine-styles="inline">.__m__-P0-2{--col-flex-grow:auto;--col-flex-basis:100%;--col-max-width:100%;}@media(min-width: 36em){.__m__-P0-2{--col-flex-grow:auto;--col-flex-basis:100%;--col-max-width:100%;}}@media(min-width: 48em){.__m__-P0-2{--col-flex-grow:auto;--col-flex-basis:50%;--col-max-width:50%;}}@media(min-width: 62em){.__m__-P0-2{--col-flex-grow:auto;--col-flex-basis:33.333333333333336%;--col-max-width:33.333333333333336%;}}</style><div class="m-96bdd299 mantine-Grid-col __m__-P0-2"><div class="m-46b77525 mantine-InputWrapper-root mantine-TextInput-root"><label data-required="true" for="mantine-P0-4" id="mantine-P0-4-label" class="m-8fdc1311 mantine-InputWrapper-label mantine-TextInput-label">FN<span aria-hidden="true" class="m-78a94662 mantine-InputWrapper-required mantine-TextInput-required"> *</span></label><p id="mantine-P0-4-description" class="m-fe47ce59 mantine-InputWrapper-description mantine-TextInput-description">Specifies the <mark>formatted</mark> text corresponding to the <mark>name</mark> of the object the vCard represents.</p><div style="--input-margin-top: calc(var(--mantine-spacing-xs) / 2);" data-variant="default" class="m-6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"><input data-variant="default" placeholder="Mr. John Q. Public, Esq." required="" aria-invalid="false" aria-describedby="mantine-P0-4-description" id="mantine-P0-4" class="m-8fb7ebe7 mantine-Input-input mantine-TextInput-input"></div></div><div class="m-46b77525 mantine-InputWrapper-root mantine-TextInput-root"><label data-required="true" for="mantine-P0-10" id="mantine-P0-10-label" class="m-8fdc1311 mantine-InputWrapper-label mantine-TextInput-label">N<span aria-hidden="true" class="m-78a94662 mantine-InputWrapper-required mantine-TextInput-required"> *</span></label><p id="mantine-P0-10-description" class="m-fe47ce59 mantine-InputWrapper-description mantine-TextInput-description">Specifies the components of the <mark>name</mark> of the object the vCard represents.</p><div style="--input-margin-top: calc(var(--mantine-spacing-xs) / 2);" data-variant="default" class="m-6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"><input data-variant="default" placeholder="Public;John;Quinlan;Mr.;Esq." required="" aria-invalid="false" aria-describedby="mantine-P0-10-description" id="mantine-P0-10" class="m-8fb7ebe7 mantine-Input-input mantine-TextInput-input"></div></div></div></div></div>"`;
